ethers-providers/src/transports/retry.rs

```diff
@@ -0,0 +1,257 @@
+//! A [JsonRpcClient] implementation that retries requests filtered by [RetryPolicy]
+//! and [Backoff] strategy.
+
+use super::http::ClientError;
+use crate::{provider::ProviderError, JsonRpcClient};
+
+use std::{
+    clone::Clone,
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Mutex,
+    },
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use backoff::ExponentialBackoff as BackoffExponentialBackoff;
+#[cfg(not(target_arch = "wasm32"))]
+use futures_timer::Delay;
+use serde::{de::DeserializeOwned, Serialize};
+use thiserror::Error;
+#[cfg(target_arch = "wasm32")]
+use wasm_timer::Delay;
+
+/// [RetryPolicy] defines logic for which [JsonRpcClient::Error] instances should
+/// the client retry the request and try to recover from.
+pub trait RetryPolicy<E> {
+    fn should_retry(&self, error: &E) -> bool;
+}
+
+/// [Backoff] defines the logic for determining the [Duration] for which
+/// the client must backoff before retrying the request.
+pub trait Backoff {
+    ///
+    fn next_backoff(&mut self, total_enqueued_requests: u32, retry_number: u64)
+        -> Option<Duration>;
+}
+
+/// [RetryClient] presents as a wrapper around [JsonRpcClient] that will retry
+/// requests based on [Backoff] strategy and filtering based on [RetryPolicy].
+#[derive(Debug)]
+pub struct RetryClient<T, U, B>
+where
+    T: JsonRpcClient,
+    T::Error: Sync + Send + 'static,
+    U: RetryPolicy<T::Error> + Send + Sync + Debug,
+    B: Backoff + Send + Sync,
+{
+    inner: T,
+    requests_enqueued: AtomicU32,
+    policy: U,
+    backoff: Mutex<B>,
```

Comment:
not a big fan of the mutex but perhaps the easiest way to make this work.
wdyt @gakonst 

nevertheless we should use the tokio::sync::Mutex here